File: backend/src/fitness_lab/domain/exercise_names.py
"""Exercise names typed by the lifter (V3.3.1 "Change" with a new exercise).

Two rules, mirrored in ``web/src/lib/exerciseNames.ts`` so the screen can show the exact name
before it is saved:

- ``typed_exercise_name``: surrounding and repeated whitespace goes; an all-lowercase entry is
  title-cased ("triceps curl" -> "Triceps Curl"), any other casing is the lifter's own and
  kept ("EZ-bar curl", "RDL").
- ``name_key``: the comparison key that keeps trivial duplicates out — case and whitespace
  insensitive ("Triceps  curl" is "Triceps Curl").
"""

from __future__ import annotations

MAX_NAME_LENGTH = 80
# A letter after one of these (or at the start) begins a word.
WORD_BREAKS = frozenset(" -/(")


def name_key(name: str) -> str:
    return " ".join(name.split()).casefold()


def typed_exercise_name(text: str) -> str:
    collapsed = " ".join(text.split())
    if not collapsed:
        raise ValueError("exercise name must not be blank")
    if len(collapsed) > MAX_NAME_LENGTH:
        raise ValueError(f"exercise name must be at most {MAX_NAME_LENGTH} characters")
    if collapsed != collapsed.lower():
        return collapsed
    characters: list[str] = []
    previous = " "
    for character in collapsed:
        characters.append(character.upper() if previous in WORD_BREAKS else character)
        previous = character
    return "".join(characters)

File: web/src/lib/exerciseNames.ts
/**
 * Exercise names typed by the lifter — the same two rules as the server's
 * backend/src/fitness_lab/domain/exercise_names.py, so the screen can show the exact name
 * before it is saved. The server stays the authority.
 */
export const MAX_NAME_LENGTH = 80
const WORD_BREAKS = new Set([' ', '-', '/', '('])

/** Case- and whitespace-insensitive comparison key: "Triceps  curl" is "Triceps Curl". */
export function nameKey(name: string): string {
  return name.trim().split(/\s+/).join(' ').toLowerCase()
}

/**
 * Surrounding and repeated whitespace goes; an all-lowercase entry is title-cased
 * ("triceps curl" → "Triceps Curl"); any other casing is the lifter's own ("EZ-bar curl").
 * Null when nothing usable was typed.
 */
export function typedExerciseName(text: string): string | null {
  const collapsed = text.trim().split(/\s+/).join(' ')
  if (collapsed === '' || collapsed.length > MAX_NAME_LENGTH) return null
  if (collapsed !== collapsed.toLowerCase()) return collapsed
  let previous = ' '
  let result = ''
  for (const character of collapsed) {
    result += WORD_BREAKS.has(previous) ? character.toUpperCase() : character
    previous = character
  }
  return result
}
